internal/metrics/metrics.go

```diff
@@ -128,3 +138,26 @@ func (ctr *StateCounter) Collect(ch chan<- prometheus.Metric) {
 	ch <- prometheus.MustNewConstMetric(ctr.desc, prometheus.GaugeValue, float64(paused), "paused")
 	ch <- prometheus.MustNewConstMetric(ctr.desc, prometheus.GaugeValue, float64(stopped), "stopped")
 }
+
+type metricsReader struct {
+	counter gometrics.Counter
+	in      io.ReadCloser
+}
+
+// NewMetricsReader returns an io.ReadCloser that counts the bytes read by
+// the provided io.ReadCloser to the provided counter.
+func NewMetricsReader(counter gometrics.Counter, in io.ReadCloser) io.ReadCloser {
+	return &metricsReader{counter: counter, in: in}
+}
```

Comment:
It's possible to get rid of the `Close()` boilerplate if you were to embed the `io.ReadCloser` into the struct. And if you were to make the `counter` field public the constructor would not be necessary, either.
```go
type MetricsReader struct {
	io.ReadCloser
	Counter gometrics.Counter
}
```